File: 1105-filling-bookcase-shelves/1105-filling-bookcase-shelves.py
class Solution:
    def __init__(self):
        self.n = 0
        self.WIDTH = 0
        self.dp = [[-1] * 1001 for _ in range(1001)]
        
    def solve(self, books, i, remain, maxHt):
        if i >= self.n:
            return maxHt
        if self.dp[i][remain] != -1:
            return self.dp[i][remain]
        
        bookW, bookH = books[i]
        keep = float('inf')
        skip = float('inf')
        
        if bookW <= remain:
            keep = self.solve(books, i + 1, remain - bookW, max(maxHt, bookH))
        skip = maxHt + self.solve(books, i + 1, self.WIDTH - bookW, bookH)
        
        self.dp[i][remain] = min(keep, skip)
        return self.dp[i][remain]
    
    def minHeightShelves(self, books: List[List[int]], shelfWidth: int) -> int:
        self.n = len(books)
        self.WIDTH = shelfWidth
        for i in range(self.n + 1):
            for j in range(1001):
                self.dp[i][j] = -1
        return self.solve(books, 0, shelfWidth, 0)

File: 1105-filling-bookcase-shelves/1105-filling-bookcase-shelves.cpp
class Solution {
public:
    int n;
    int WIDTH;
    int dp[1001][1001];
    int solve(vector<vector<int>> &books, int i, int remain, int maxHt) {
        if (i >= n) {
            return maxHt;
        }
        if (dp[i][remain] != -1) {
            return dp[i][remain];
        }
        int bookW = books[i][0];
        int bookH = books[i][1];
        int keep = INT_MAX;
        int skip = INT_MAX;
        if (bookW <= remain) {
            keep = solve(books, i+1, remain - bookW, max(maxHt, bookH));
        }
        skip = maxHt + solve(books, i+1, WIDTH - bookW, bookH);
        return dp[i][remain] = min(keep, skip);

    }
    int minHeightShelves(vector<vector<int>>& books, int shelfWidth) {
        n = books.size();
        WIDTH = shelfWidth;
        memset(dp, -1, sizeof(dp));
        int reaminingWidth = shelfWidth;
        return solve(books, 0, reaminingWidth, 0);
    }
};